src/actions/EditNewLine.ts

```diff
@@ -1,66 +1,139 @@
     this.run = this.run.bind(this);
   }
 
-  private correctForParagraph(targets: Target[]) {
-    targets.forEach((target) => {
-      let { start, end } = target.contentRange;
-      if (target.scopeType === "paragraph") {
-        if (this.isAbove && target.leadingDelimiter != null) {
-          start = start.translate({ lineDelta: -1 });
-        } else if (!this.isAbove && target.trailingDelimiter != null) {
-          end = end.translate({ lineDelta: 1 });
-        }
-        target.contentRange = new Range(start, end);
-      }
-    });
-  }
+  async run([targets]: [Target[]]): Promise<ActionReturnValue> {
+    const editor = ensureSingleEditor(targets);
 
-  private isNotebookEditor(editor: TextEditor) {
-    return getNotebookFromCellDocument(editor.document) != null;
-  }
+    const targetsWithContext = targets.map((target) => ({
+      target,
+      context: target.getEditNewLineContext(this.isBefore),
+    }));
+    const commandTargets = targetsWithContext.filter(
+      ({ context }) => !!(<any>context).command
+    );
+    const delimiterTargets = targetsWithContext.filter(
+      ({ context }) => !!(<any>context).delimiter
+    );
 
-  private getCommand(target: Target) {
-    if (target.scopeType === "notebookCell") {
-      if (this.isNotebookEditor(target.editor)) {
-        return this.isAbove
-          ? "notebook.cell.insertCodeCellAbove"
-          : "notebook.cell.insertCodeCellBelow";
-      }
-      return this.isAbove
-        ? "jupyter.insertCellAbove"
-        : "jupyter.insertCellBelow";
+    if (commandTargets.length > 0 && delimiterTargets.length > 0) {
+      throw new Error("Can't insert edit using command and delimiter at once");
+    }
+
+    if (commandTargets.length > 0) {
+      const commands = commandTargets.map(
+        ({ context }) => (<any>context).command
+      );
+      return {
+        thatMark: await this.runCommand(targets, commands),
+      };
     }
-    return this.isAbove
-      ? "editor.action.insertLineBefore"
-      : "editor.action.insertLineAfter";
+
+    return {
+      thatMark: await this.runDelimiter(targets, editor),
+    };
   }
 
-  async run([targets]: [Target[]]): Promise<ActionReturnValue> {
-    this.correctForParagraph(targets);
+  async runDelimiter(targets: Target[], editor: TextEditor) {
+    const edits = targets.map((target) => {
+      const { contentRange } = target;
+      const context = target.getEditNewLineContext(this.isBefore);
+      const delimiter = (<any>context).delimiter as string;
 
-    if (this.isAbove) {
+      // Delimiter is one or more new lines. Handle as lines.
+      if (delimiter.includes("\n")) {
+        const lineNumber = this.isBefore
+          ? contentRange.start.line
+          : contentRange.end.line;
+        const line = editor.document.lineAt(lineNumber);
+        const characterIndex = line.isEmptyOrWhitespace
+          ? contentRange.start.character
+          : line.firstNonWhitespaceCharacterIndex;
```

Comment:
This will always just be same indentation as previous line, right?  The built-in VSCode indentation code will automatically indent if, for example, you're on the guard of an `if` statement.  I don't think we can recover this behaviour without running the VSCode command or doing something like https://github.com/cursorless-dev/cursorless/issues/685

Reply:
Correct. Look in comment 2. We could just default to existing vscode command and only use delimiter for other than single line if you like.

Reply:
I think I would just override the context in the Line class, the way you've done for notebook cells